ops/pebble.py

```diff
@@ -0,0 +1,487 @@
+# Copyright 2019-2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Client for the Pebble API (HTTP over Unix socket)."""
+
+# TODO(benhoyt): consider the following:
+# - think about how we'll handle API versioning
+# - add automatic retries
+# - unify errors into package-local error
+
+from typing import Dict, List, Optional
+import datetime
+import enum
+import http.client
+import json
+import os
+import re
+import socket
+import time
+import urllib.error
+import urllib.parse
+import urllib.request
+
+
+class _UnixSocketConnection(http.client.HTTPConnection):
+    """Implementation of HTTPConnection that connects to a named Unix socket."""
+
+    def __init__(self, host, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, socket_path=None):
+        super(_UnixSocketConnection, self).__init__(host, timeout=timeout)
+        self.socket_path = socket_path
+
+    def connect(self):
+        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        self.sock.connect(self.socket_path)
+        if self.timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+            self.sock.settimeout(self.timeout)
+
+
+class _UnixSocketHandler(urllib.request.AbstractHTTPHandler):
+    """Implementation of HTTPHandler that uses a named Unix socket."""
+
+    def __init__(self, socket_path):
+        super(_UnixSocketHandler, self).__init__()
+        self.socket_path = socket_path
+
+    def http_open(self, req):
+        return self.do_open(_UnixSocketConnection, req, socket_path=self.socket_path)
+
+    http_request = urllib.request.AbstractHTTPHandler.do_request_
+
+
+_TIMESTAMP_RE = re.compile(r'(.*)\.(\d+)(.*)')
+
+
+def _parse_timestamp(s):
+    """Parse timestamp from Go-encoded JSON (which uses 9 decimal places for seconds."""
+    match = _TIMESTAMP_RE.match(s)
```

Comment:
Yeah, the `_TIMESTAMP_RE` is very loose, but we're passing the result through `fromisoformat()` to do the hard work in both cases, so I think that's okay.

Reply:
Partly to simplify the logic, but partly because strptime's `%z` handling was changed in Python 3.7 (and we need to support Python 3.5+), I've modified this to use a fairly tight regex that only matches what we want, then canonicalizes it and feeds to strptime.